### ISSUE TITLE
speaker diarization

### ISSUE THREAD

**Issue by devSJR** [mkiol/dsnote#84]
Dsnote is great for STT by using whisper. For audio samples with different persons speaking, e.g. podcasts, movies …, one ends up with a messy text because Whisper doesn’t do what’s called ‘speaker diarization’. That is, identifying one voice or another. It seems there is a solution to this. Maybe you want to check the following:

https://ultracrepidarian.phfactor.net/

**Comment by mkiol** [comment]
Definitely, looks very interesting. 

Processing pipeline seems to be as follows:
1. Audio transcription => "words" + timestamps
2. Audio segmentation => "speaker-id" + timestamps
3. Matching "words" to "speaker-id" based on timestamps

Currently Speech Note does not recognize timestamps, so this has to be implemented, but I've already need timestamps for [subtitles support](https://github.com/mkiol/dsnote/issues/55).

"Audio segmentation" is done with extra model [pyannote/segmentation-3.0](https://huggingface.co/pyannote/segmentation-3.0). To download this model you need to have Hugging Face account which might be problematic.

I will investigate what can be done.

**Comment by devSJR** [comment]
I am not really in need of this, but can imagine that other users might want this.
I guess eventually there will be an easily accessible open-source audio segmentation model.
Actually, it is quite surprising that there is so much available for users free of charge (including giving some personal information like an e-mail-address).

**Comment by thob** [comment]
I'm actually hoping for diarization. my use case (discourse analytics) benefits from a stable differentiation.

**Comment by mkiol** [comment]
I did some research to find out what is possible. It looks as follows:

- Almost everyone uses [pyannote](https://huggingface.co/pyannote) segmentation models for diarization. The models work well... but not perfectly. The main problem is that the models are made available under the MIT license, but to download them you have to agree to certain conditions. To download them from HuggingFace you have to have an account and agree to the following conditions:

![image](https://github.com/mkiol/dsnote/assets/1420902/8549a0c0-e581-43e8-b47f-d48d7fd2618e)

I must say, I don't like it. Especially this "You need to agree to share your contact". Speech Note is a privacy focused application and "sharing your contact information" doesn't fit well.

- There is also “experimental” support for diarization in [whisper.cpp](https://github.com/ggerganov/whisper.cpp#speaker-segmentation-via-tinydiarize-experimental). I love `whisper.cpp` and it's already integrated into Speech Note. The problem is that to use diarization you need to download a special model that combines STT and diarization. Currently this single model is only for English :/

I'm keep looking for another better solution...

**Comment by devSJR** [comment]
You are right. Privacy is an important asset.
Regarding the  “experimental” support for diarization in whisper.cpp, I think you should give it a go, if it is not too difficult to implement, and even if it is only for English at the moment.

**Comment by thob** [comment]
I'd need it for German … I checked the thread above, from what I understand it's still rather experimental. As a user I would be OK to accept the terms for being able to diarize  _– as long as there's no viable alternative._

**Comment by thob** [comment]
I found this repo, maybe it could be leveraged somehow https://github.com/m-bain/whisperX but it seems to rely on pyannote as well.

**Comment by mkiol** [comment]
Yes, WhisperX uses the same pyannote models. Therefore you have to [pass HF token](https://github.com/m-bain/whisperX#speaker-diarization) to use diarization :(

**Comment by devSJR** [comment]
Just learned about this

https://joss.theoj.org/papers/10.21105/joss.05266

Diart: A Python Library for Real-Time Speaker
Diarization

**Comment by mkiol** [comment]
@devSJR unfortunately same pyannote models are needed to make it work :(

https://github.com/juanmc2005/diart?tab=readme-ov-file#get-access-to--pyannote-models

**Comment by devSJR** [comment]
OK, will keep looking

**Comment by machiav3lli** [comment]
Possibly relevant sources (some are more directly usable than others):
- [SpeechBrain](https://github.com/speechbrain/speechbrain)
- [Awesome-Diarization](https://github.com/wq2012/awesome-diarization)
- [Resemblyzer](https://github.com/resemble-ai/Resemblyzer)
- [Kaldi](https://github.com/cadia-lvl/kaldi-speaker-diarization)

**Comment by devSJR** [comment]
I have just learned about https://github.com/bugbakery/transcribee/pkgs/container/transcribee 

In transcribee they seem to be able to attribute text to a speaker. Maybe worth a look?

**Comment by machiav3lli** [comment]
There's now also https://github.com/thewh1teagle/vibe with speaker-diarization baked-in

**Comment by maurerle** [comment]
vibe and transcribee also rely on a HUGGINGFACE_ACCESS_TOKEN and pyannote/segmentation-3.0.onnx

**Comment by m13v** [comment]
Speaker diarization would be a great addition to dsnote. We build Fazm, a macOS AI agent that uses WhisperKit for on-device transcription, and adding speaker identification was one of the most requested features.

A few practical notes from our implementation:

1. **pyannote.audio** is probably the most mature open-source option for diarization. It works well as a post-processing step after Whisper transcription -- you run Whisper first for the text + timestamps, then pyannote on the same audio to get speaker segments, and merge the two outputs by aligning timestamps. The quality is solid for 2-4 speakers.

2. **Performance consideration**: diarization adds meaningful processing time. For a 1-hour podcast, expect an additional 2-5 minutes depending on hardware. It might be worth making it an opt-in toggle rather than running it by default, so users who just want quick transcription are not slowed down.

3. **Output format**: labeling speakers as "Speaker 1", "Speaker 2" etc. with a way to rename them after the fact works well. Some users know who is speaking and want to label them; others just want the separation.

4. **whisperX** (https://github.com/m-bain/whisperX) combines Whisper with forced alignment and diarization in a single pipeline, which could be a reference for how to integrate it cleanly.

The link you shared about the ultracrepidarian approach is interesting too -- would love to see diarization support land in dsnote.

**Comment by m13v** [comment]
Here is our WhisperKit-based transcription service for reference on how we integrated on-device speech recognition: https://github.com/m13v/fazm/blob/main/Desktop/Sources/TranscriptionService.swift -- the timestamp alignment approach might be relevant when integrating diarization output with Whisper transcripts.
